content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+
+For example, every formula has what is called a [“Disjunctive Normal Form”
+(DNF)](https://en.wikipedia.org/wiki/Disjunctive_normal_form), so called
+because it consists of a series of conjunction (“and”) terms, joined together
+by disjunction (“or”) operators[^3]:
+
+    (X ∧ Z) ∨ (¬Y ∧ Z)
+
+Each term has a subset of the variables, possibly negated, joined by ∧. The
+terms are then joined together using `∨`.
+
+Solving the satisfiability problem for a formula in DNF is obviously very easy:
+
+1. Go through the individual terms. `∨` is true if and only if either of its
+   operands is true, so for each term:
+   * If it contains both a variable and its negation (i.e. `X∧¬X`) it can
+     never be true, continue to the next term.
+   * Otherwise, you can read an assignment from the term - assign `true`, if
+     the term contains `X` and assign `false`, if it contains `¬X`. The term
+     is then true, so the entire formula is.
+2. If none of the terms can be made `true`, the formula can't be made `true`
+   and there is no solution.
+
+On the other hand, there is also a [“Conjunctive Normal
+Form”](https://en.wikipedia.org/wiki/Conjunctive_normal_form). Here, the
+formula is a series of *disjunction* terms, joined together with a
+*conjunction*:
+
+    (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+
+For this, the same idea does not work - you have to take *all terms* into
+account simultaneously, to find an assignment and can't just tackle them one by
+one. In fact, solving satisfiability on CNF (often abbreviated as “CNFSAT”) is
+NP-complete[^4].
+
+Very often, when we want to prove a problem is hard, we do so by reducing it to
+CNFSAT. That's what we will do for the problem of calculating type sets. But
+there is one more preamble we need.
+
+## Sets and Satisfiability
+
+There is an important relationship between
+[sets](https://en.wikipedia.org/wiki/Set_(mathematics)) and boolean functions
+(functions which return `true` or `false`).
+
+Say we have some universe of objects `U`. And say we have a boolean function
+`f:U→B` with `B:={true,false}`. We can then create a set from that, by looking
+at all objects for which `f` is `true`:
+
+    S[f] := { x∈U | f(x) }
+
+We use the index form `S[f]` to indicate, that `S` depends on the function `f`.
+
+Likewise, if we have any set `S⊆U`, we can create a boolean function from it:
+
+    f[S]: U → B
+          x ↦ x∈S
+
+That is, the function returns `true` if an element is in `S` and `false`
+otherwise.
+
+An interesting aspect of this, is that we can translate between boolean
+operators and set operators this way. For example, set union `∪` becomes
+equivalent to logical disjunction `∨`:
+
+    S[f] ∪ S[g] = { x∈U | f(x) } ∪ { x∈U | g(x) }
+                = { x∈U | f(x) ∨ g(x) }
+
+    f[S](x) ∨ f[T](x) = x∈S ∨ x∈T = x∈(S∪T)
+
+Similarly, `∩` becomes equivalent to `∧`, `⊆` becomes equivalent to `⇒` and `¬`
+becomes equivalent to set complement `U\S := {x∈U | x∉S}`.
```

Comment:
Bad habits die hard. I'm used to just throw in a definition, if I haven't used an operator before and have people know what I mean. That is, the `:=…` part is "explaining what `\` is" :)

I had some trouble when writing this, because there is no well-known single symbol for "set complement", I think.

Changed this, let me know if you have any better ideas (once I pushed).